src/sparse_blas/backends/mkl_common/mkl_handles.cxx

```diff
@@ -0,0 +1,393 @@
+/***************************************************************************
+*  Copyright (C) Codeplay Software Limited
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  For your convenience, a copy of the License has been included in this
+*  repository.
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+*
+**************************************************************************/
+
+// Dense vector
+template <typename fpType>
+void init_dense_vector(sycl::queue & /*queue*/,
+                       oneapi::mkl::sparse::dense_vector_handle_t *p_dvhandle, std::int64_t size,
+                       sycl::buffer<fpType, 1> val) {
+    *p_dvhandle = new oneapi::mkl::sparse::dense_vector_handle(val, size);
+}
+
+template <typename fpType>
+void init_dense_vector(sycl::queue & /*queue*/,
+                       oneapi::mkl::sparse::dense_vector_handle_t *p_dvhandle, std::int64_t size,
+                       fpType *val) {
+    *p_dvhandle = new oneapi::mkl::sparse::dense_vector_handle(val, size);
+}
+
+template <typename fpType, typename InternalHandleT>
+void check_can_reset_value_handle(const std::string &function_name,
+                                  InternalHandleT *internal_handle, bool expect_buffer) {
+    if (internal_handle->get_value_type() != detail::get_data_type<fpType>()) {
+        throw oneapi::mkl::invalid_argument(
+            "sparse_blas", function_name,
+            "Incompatible data types expected " +
+                data_type_to_str(internal_handle->get_value_type()) + " but got " +
+                data_type_to_str(detail::get_data_type<fpType>()));
+    }
+    if (internal_handle->all_use_buffer() != expect_buffer) {
+        throw oneapi::mkl::invalid_argument(
+            "sparse_blas", function_name, "Cannot change the container type between buffer or USM");
+    }
+}
+
+template <typename fpType>
+void set_dense_vector_data(sycl::queue & /*queue*/,
+                           oneapi::mkl::sparse::dense_vector_handle_t dvhandle, std::int64_t size,
+                           sycl::buffer<fpType, 1> val) {
+    check_can_reset_value_handle<fpType>(__FUNCTION__, dvhandle, true);
+    dvhandle->size = size;
+    dvhandle->set_buffer(val);
+}
+
+template <typename fpType>
+void set_dense_vector_data(sycl::queue & /*queue*/,
+                           oneapi::mkl::sparse::dense_vector_handle_t dvhandle, std::int64_t size,
+                           fpType *val) {
+    check_can_reset_value_handle<fpType>(__FUNCTION__, dvhandle, false);
```

Comment:
Aren't we supposed to use `__func__` since that's part of the standard? I don't know if `__FUNCTION__` is part of C/C++ standards. E.g.: https://stackoverflow.com/questions/4384765/whats-the-difference-between-pretty-function-function-func

Reply:
Done in https://github.com/oneapi-src/oneMKL/pull/500/commits/9dbd67bde71a9a2b19905509cf2d63117d0e87b7